oligo_designer_toolsuite/oligo_specificity_filter/_filter_blastn.py

```diff
@@ -187,15 +217,186 @@ def _find_hits(
         if consider_hits_from_input_region:
             # remove all hits where query and reference come from the same region
             search_results = search_results[
-                search_results["query_region_id"] != search_results["reference_region_id"]
+                search_results["query_region_id"]
+                != search_results["reference_region_id"]
             ]
 
         blast_table_hits = search_results.loc[
             search_results.alignment_length > search_results.min_alignment_length
         ]
-        oligos_with_hits = blast_table_hits["query"].unique()
 
-        return blast_table_hits, oligos_with_hits
+        return blast_table_hits
+
+    def _get_references(
+        self, table_hits: pd.DataFrame, file_reference: str, region_id: str
+    ):
+        """
+        Retrieve the reference sequences from the search results.
+
+        :param table_hits: Dataframe containing the search results.
+        :type searchtable_hits_results: pd.DataFrame
+        :param file_reference: Path to the fasta file used as reference for the search.
+        :type file_reference: str
+        :param region_id: The identifier for the region within the database to filter.
+        :type region_id: str
+        :return: Reference sequences
+        :rtype: list
+        """
+        required_fields = [
+            "query",
+            "reference",
+            "alignment_length",
+            "query_start",
+            "query_end",
+            "query_length",
+            "query_sequence",
+            "reference_start",
+            "reference_end",
+            "reference_sequence",
+            "reference_strand"
+        ]
+        if not all(field in self.names_search_output for field in required_fields):
+            raise ValueError(
+                f"Some of the required fields {required_fields} are missing in the search results."
+            )
+        
+        # set the positions to a 0-based index
+        references_plus = table_hits[table_hits["reference_strand"] == "plus"]
+        references_plus.loc[:, "query_start"] = references_plus["query_start"] - 1
+        references_plus.loc[:, "reference_start"] = references_plus["reference_start"] - 1
+        references_minus = table_hits[table_hits["reference_strand"] == "minus"]
+        references_minus.loc[:, "query_start"] = references_minus["query_start"] - 1
+        references_minus.loc[:, "reference_end"] = references_minus["reference_end"] - 1
+        table_hits = pd.concat([references_plus, references_minus])
+        table_hits.sort_index(inplace=True)  # restore the intial indexes
+
```

Comment:
consider optimizing the code above to avoid slow slicing and concat:
```
table_hits["query_start"] = table_hits["query_start"] - 1
table_hits["reference_start"] = np.where(table_hits["reference_strand"] == "plus", table_hits["reference_start"] - 1, table_hits["reference_start"])
table_hits["reference_end"] = np.where(table_hits["reference_strand"] == "minus", table_hits["reference_end"] - 1, table_hits["reference_end"])

```